### PR TITLE
Fixing signing issues for powershell

### PR DESCRIPTION
**Pull request by shahabhijeet**
1. Addressing Signing bugs. Now we sign all dll with Sha2 with Sha1, MSI with Sha1, all scripts with Sha2.
2. Removing all .js as well as json files from layouts.
3. Adding Setup tests. Regenerating wix files.
4. Fixing reference for ASM-ExpressRoute causing it to be part of the layout.
5. Removing iisnode.dll, iisnode_schema.xml from layout

### CODE REVIEW

File: AzurePowershell.Test.targets [hunk 1]
```diff
@@ -1,531 +1,586 @@
...
+		</ItemGroup>
+		<ItemGroup Condition=" '$(scope)' == 'ServiceManagement' ">
+			<XUnitTests Include="@(AsmXUnitTests)"/>
+		</ItemGroup>
+		<ItemGroup Condition=" '$(scope)' != 'all' and '$(scope)' != 'ServiceManagement' and '$(scope)' != 'AzureStorage' ">
+			<XUnitTests Include="$(MSBuildProjectDirectory)\src\ResourceManager\$(scope)\*\bin\Debug\*.Test.dll"/>
+		</ItemGroup>
+		<ItemGroup Condition=" '$(Configuration)' == 'Release' AND '$(scope)' == 'all' AND '$(CodeSign)' == 'true' ">
+			<!-- We want to run setup tests related to signing that can only be run on signed bits -->
+			<XUnitRelTests Include=".\setup\PowerShellSetup.Test\bin\Release\PowerShellSetup.Test.dll" />
```
Comment (hovsepm) [inline AzurePowershell.Test.targets]: >XUnitRelTests [](start = 4, length = 13)

the name is confusing. what about XUnitSignedTests?

File: AzurePowershell.Test.targets [hunk 2]
```diff
@@ -1,531 +1,586 @@
...
+		<Message Importance="high" Text="Setup XUnit tests" />
+		<MakeDir Directories="$(TestOutputDirectory)" ContinueOnError="false" />
+		
+		<ItemGroup Condition=" '$(scope)' == 'all' AND '$(CodeSign)' != 'true' ">
+			<SetupDebugTests Include=".\setup\PowerShellSetup.Test\bin\Debug\PowerShellSetup.Test.dll" />
+		</ItemGroup>
+		
+		<ItemGroup Condition=" '$(Configuration)' == 'Release' AND '$(scope)' == 'all' AND '$(CodeSign)' == 'true' ">
+			<!-- We want to run setup tests related to signing that can only be run on signed bits -->
+			<SetupRelTests Include=".\setup\PowerShellSetup.Test\bin\Release\PowerShellSetup.Test.dll" />
```
Comment (hovsepm) [inline AzurePowershell.Test.targets]: >SetupRelTests [](start = 4, length = 13)

same comment - what does Rel even mean?

File: build.proj [hunk 3]
```diff
@@ -144,61 +155,73 @@
...
       Properties="Configuration=Debug;Platform=Any CPU"
       ContinueOnError="false" />
 
-    <CallTarget Targets="DependencyAnalysis" />
-    
-    <CallTarget Targets="CodeSignInstaller" 
-                Condition=" '$(CodeSign)' == 'true' and '$(Scope)' == 'all'" />    
+    <CallTarget targets="DependencyAnalysis" />
+
+	<CallTarget Targets="BuildSetup"/>
```
Comment (hovsepm) [inline build.proj]: >Targets [](start = 13, length = 7)

out of curiosity, why did you change casing in "Targets" above?

File: setup/azurecmdfiles.wxi [hunk 4]
```diff
@@ -4,6 +4,62 @@
         <DirectoryRef Id="PowerShellFolder">
             <Directory Id="dirB60B5B1F56FA5598EFFC960CA393FDCE" Name="ResourceManager">
                 <Directory Id="dir9E409EB63EF03736275B51044C819FD8" Name="AzureResourceManager">
+                    <Directory Id="dirE5EE64AF112D660B2F38178FA55F7183" Name="Azure.AnalysisServices">
```
Comment (hovsepm) [inline setup/azurecmdfiles.wxi]: >dirE5EE64AF112D660B2F38178FA55F7183 [](start = 35, length = 35)

Why do you have these changes? they does not look related at all to "Fixing signing issues for powershell".

**Comment by hovsepm** [comment]
> namespace PowerShellSetup.Tests

please add license header


---
Refers to: setup/PowerShellSetup.Test/MsiLayoutTests.cs:1 in 8681907. [](commit_id = 868190772286b67c69c15348c58596d54269b9cd, deletion_comment = False)

**Comment by hovsepm** [comment]
> namespace PowerShellSetup.Tests

license header missing

---
Refers to: setup/PowerShellSetup.Test/MsiTestBase.cs:1 in 8681907. [](commit_id = 868190772286b67c69c15348c58596d54269b9cd, deletion_comment = False)

**Comment by hovsepm** [comment]
> using System.Reflection;

License header missing

---
Refers to: setup/PowerShellSetup.Test/Properties/AssemblyInfo.cs:1 in 8681907. [](commit_id = 868190772286b67c69c15348c58596d54269b9cd, deletion_comment = False)

File: AzurePowershell.Test.targets (hunk 1 above)
Comment (shahabhijeet) [inline AzurePowershell.Test.targets]: @hovsepm in order to run tests for signed jobs, it needs to be build as part of Release configuration.
Similar to XUnitDebugTests, I thought of XUnitRelTests, I can make it XUnitReleaseTests.
Dbg and Rel are commonly used so thought of using that instead of XUnitReleaseTests.

File: AzurePowershell.Test.targets (hunk 2 above)
Comment (shahabhijeet) [inline AzurePowershell.Test.targets]: @hovsepm similar to above comment.
Rel = Release

File: build.proj (hunk 3 above)
Comment (shahabhijeet) [inline build.proj]: @hovsepm I did not mean to, but it was mistake on my part.
I had commented that line earlier during testing.
VS has a shortcut to uncomment Ctrl+uk, but TextPad++ has the same keyboard shortcut to make everything lowercase and I ended up checking that without doing due diligence.

File: setup/azurecmdfiles.wxi (hunk 4 above)
Comment (shahabhijeet) [inline setup/azurecmdfiles.wxi]: @hovsepm the entire regen need to happen because I removed .js and .json files from the two remaining RPs that still had them. And caused the entire regen.

File: tools/Installer/generate.ps1 [hunk 5]
```diff
@@ -25,6 +25,14 @@ if ([string]::IsNullOrEmpty($buildConfig))
 	$buildConfig = 'Release'
 }
 
+#if((Test-Path -Path $env:AzurePSRoot) -eq $false)
+if($env:AzurePSRoot -eq $null)
+{
+    $env:AzurePSRoot="..\..\"
```
Comment (markcowl) [inline tools/Installer/generate.ps1]: Can't we set this to $PSScriptRoot\..\..\ ?

File: tools/Installer/generate.ps1 [hunk 6]
```diff
@@ -42,11 +50,17 @@ Remove-Item -Force $resourceManagerPath\AzureRM.DataLakeStore\Microsoft.Azure.Co
...
 Remove-Item -Force $resourceManagerPath\AzureRM.RecoveryServices.Backup\AzureRM.RecoveryServices.psd1 -ErrorAction SilentlyContinue
+Remove-Item -Recurse -Force $resourceManagerPath\Azure.AnalysisServices\ -ErrorAction SilentlyContinue
+
 Write-Verbose "Removing duplicated Resources folder"
 Remove-Item -Recurse -Force $serviceManagementPath\Compute\Resources\ -ErrorAction SilentlyContinue
 Remove-Item -Recurse -Force $serviceManagementPath\Sql\Resources\ -ErrorAction SilentlyContinue
 Remove-Item -Recurse -Force $serviceManagementPath\Storage\Resources\ -ErrorAction SilentlyContinue
 Remove-Item -Recurse -Force $serviceManagementPath\ManagedCache\Resources\ -ErrorAction SilentlyContinue
+Remove-Item -Recurse -Force $serviceManagementPath\Networking\Resources\ -ErrorAction SilentlyContinue
+Remove-Item -Recurse -Force $serviceManagementPath\Services\Resources\ -ErrorAction SilentlyContinue
```
Comment (markcowl) [inline tools/Installer/generate.ps1]: These should not be removed

Comment (markcowl) [inline tools/Installer/generate.ps1]: Also, why delete Services\Resources twice?

File: setup/bin/iisnode_schema.xml [hunk 7]
```diff
@@ -1,58 +0,0 @@
-<!--
```
Comment (markcowl) [inline setup/bin/iisnode_schema.xml]: Why remove?

Comment (markcowl) [inline setup/bin/iisnode_schema.xml]: iisnode is part of the hosting environment for ServiceProject apps

File: AzurePowershell.Test.targets [hunk 8]
```diff
@@ -1,531 +1,586 @@
...
-    <LiveTestDlls Include="$(ResourceManagerTestDebug)" />    
-  </ItemGroup>
-
-  <Target Name="LiveTests"> 
-    <MakeDir Directories="$(TestOutputDirectory)" ContinueOnError="false" />
-    <Exec Command="packages\xunit.runner.console.2.1.0\tools\xunit.console.x86.exe  @(LiveTestDlls) /trait &quot;AcceptanceType=LiveBVT&quot; /html &quot;$(TestOutputDirectory)\Live.%(Filename).html&quot;" />
-  </Target> 
-
-
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="4.0">
```
Comment (shahabhijeet) [inline AzurePowershell.Test.targets]: @markcowl I actually fixed the tabs in the entire file.

File: build.proj [hunk 9]
```diff
@@ -1,30 +1,30 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  
+
```
Comment (shahabhijeet) [inline build.proj]: @markcowl yes because in last couple of commits, the tabs are not set right in our file, so formatted the entire file.

File: setup/PowerShellSetup.Test/MsiLayoutTests.cs [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    /// </summary>
+    public class MsiLayoutTests: MsiTestBase
+    {
+        const string MSI_NAME = "AzurePowerShell.msi";
+        const string MSI_EXTRACT_DIR_NAME = "msiContents";
+        string _procOutput;
+        string _procErr;
+
+        public MsiLayoutTests(ITestOutputHelper testOutput) : base(testOutput)
+        {
```
Comment (shahabhijeet) [inline setup/PowerShellSetup.Test/MsiLayoutTests.cs]: @markcowl in tasks, we filter files from our output, but we still add change files before we package it into msi. So the best verification would be to see what is signed and not signed inside the msi.
Every test extracts it once prior to running test.

File: setup/PowerShellSetup.Test/MsiLayoutTests.cs [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+            _procOutput = string.Empty;
+            _procErr = string.Empty;
+        }
+
+        [Fact]
+        [Trait("AcceptanceType", "CheckIn")]
```
Comment (shahabhijeet) [inline setup/PowerShellSetup.Test/MsiLayoutTests.cs]: @markcowl avoiding as little dependency as possible, because we run all our tests as debug, so when we run in Release, we start getting loading errors, so avoided whatever dependency I could avoid.

File: setup/PowerShellSetup.Test/MsiLayoutTests.cs [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+        {
+            string SHA1 = "sha1RSA";
+            string SHA2 = "sha256RSA";
+
+            List<string> expectedSignatureAlgos = new List<string>() { SHA1, SHA2 };
+            string msiContentsDir = this.ExtractMsiContents(out _procErr);
+            Assert.True(Directory.Exists(msiContentsDir));
+            Assert.True(string.IsNullOrEmpty(_procErr));
+
+            IEnumerable<string> msiFiles = Directory.EnumerateFiles(msiContentsDir, "*", SearchOption.AllDirectories);
```
Comment (shahabhijeet) [inline setup/PowerShellSetup.Test/MsiLayoutTests.cs]: @markcowl the primary reason to do this is to find out whatever is laid down on user machine from the msi is properly signed. We can change our build system and might include unsigned files, so this will insure we have all properly signed files.

File: setup/azurecmdfiles.wxi [hunk 13]
```diff
@@ -5064,219 +4922,6 @@
                         <Component Id="cmpE9EDC1586045D7CEED49D3E56F77F331" Guid="*">
                             <File Id="fil4C98AB797C284E8FA2342CD4B2651CAE" KeyPath="yes" Source="$(var.sourceDir)\ServiceManagement\Azure\Services\System.Spatial.dll" />
                         </Component>
-                        <Directory Id="dir7B3D975F9D09F45FF8AC88A91847AAAE" Name="Resources">
```
Comment (shahabhijeet) [inline setup/azurecmdfiles.wxi]: @markcowl under what scenario do we use these scaffolding resources?

File: tools/Installer/generate.ps1 (hunk 6 above)
Comment (shahabhijeet) [inline tools/Installer/generate.ps1]: @markcowl what line # do you see it getting deleted twice? I am not able to find it if it was deleted earlier in the build or some place else?

File: setup/bin/iisnode_schema.xml (hunk 7 above)
Comment (shahabhijeet) [inline setup/bin/iisnode_schema.xml]: @markcowl could point me to the scenario or code that uses these binaries.

File: tools/Installer/generate.ps1 (hunk 5 above)
Comment (shahabhijeet) [inline tools/Installer/generate.ps1]: Done

File: setup/PowerShellSetup.Test/MsiLayoutTests.cs [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+                        }
+                        break;
+                    }
+                }
+            }
+
+            return friendlyAlgorithmName;
+        }
+
+        //private void CheckFileSignature()
```
Comment (shahabhijeet) [inline setup/PowerShellSetup.Test/MsiLayoutTests.cs]: Done

**Comment by shahabhijeet** [comment]
Passed setup job
http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/APS-SetupTests/

File: setup/bin/iisnode_schema.xml (hunk 7 above)
Comment (shahabhijeet) [inline setup/bin/iisnode_schema.xml]: @markcowl this issue has been resolved. We do not need these files any more.

File: setup/azurecmdfiles.wxi [hunk 15]
```diff
@@ -4501,219 +4572,6 @@
                         <Component Id="cmp1E3DC7429C181CAEFC59EB63C0ED829D" Guid="*">
                             <File Id="fil3547DC5AD472CB45B042747FB62AD147" KeyPath="yes" Source="$(var.sourceDir)\ServiceManagement\Azure\Networking\System.Spatial.dll" />
                         </Component>
-                        <Directory Id="dir02933E12768B8CBD63FAC64CED66D9DB" Name="Resources">
```
Comment (shahabhijeet) [inline setup/azurecmdfiles.wxi]: @markcowl this is fixed now.
